File: frontend/src/services/detections.ts
import { apiClient } from '@/lib/api';
import { Detection } from '@/types';

export interface DetectionListResponse {
  items: Detection[];
  meta: {
    total: number;
    page: number;
    page_size: number;
    total_pages: number;
  };
}

export const detectionService = {
  async scanCrop(
    file: File,
    options?: { farm_id?: string; field_id?: string; crop_id?: string }
  ): Promise<Detection> {
    const formData = new FormData();
    formData.append('file', file);
    if (options?.farm_id) formData.append('farm_id', options.farm_id);
    if (options?.field_id) formData.append('field_id', options.field_id);
    if (options?.crop_id) formData.append('crop_id', options.crop_id);

    return apiClient<Detection>('/detections/scan', {
      method: 'POST',
      body: formData,
    });
  },

  async listDetections(params: {
    page?: number;
    page_size?: number;
    crop_id?: string;
    farm_id?: string;
    field_id?: string;
    detection_type?: string;
    severity?: string;
    risk_level?: string;
    expert_verified?: boolean;
    search?: string;
    sort_by?: string;
  } = {}): Promise<DetectionListResponse> {
    const query = new URLSearchParams();
    if (params.page) query.append('page', params.page.toString());
    if (params.page_size) query.append('page_size', params.page_size.toString());
    if (params.crop_id) query.append('crop_id', params.crop_id);
    if (params.farm_id) query.append('farm_id', params.farm_id);
    if (params.field_id) query.append('field_id', params.field_id);
    if (params.detection_type) query.append('detection_type', params.detection_type);
    if (params.severity) query.append('severity', params.severity);
    if (params.risk_level) query.append('risk_level', params.risk_level);
    if (params.expert_verified !== undefined) query.append('expert_verified', params.expert_verified.toString());
    if (params.search) query.append('search', params.search);
    if (params.sort_by) query.append('sort_by', params.sort_by);

    return apiClient<DetectionListResponse>(`/detections?${query.toString()}`);
  },

  async getDetectionDetail(detectionId: string): Promise<Detection> {
    return apiClient<Detection>(`/detections/${detectionId}`);
  },
};

File: backend/app/api/v1/detections.py
from fastapi import APIRouter, Depends, UploadFile, File, Form, Query, status
from sqlalchemy.ext.asyncio import AsyncSession
from typing import Optional
from app.core.database import get_db
from app.core.response import success_response
from app.schemas.detection import DetectionFilterParams, DetectionResponse
from app.services.detection_service import detection_service
from app.api.v1.deps import get_current_user
from app.models.user import User

router = APIRouter(prefix="/detections", tags=["Crop Disease & Pest Detections"])

@router.post("/scan", response_model=dict, status_code=status.HTTP_201_CREATED)
async def scan_crop(
    file: UploadFile = File(...),
    farm_id: Optional[str] = Form(None),
    field_id: Optional[str] = Form(None),
    crop_id: Optional[str] = Form(None),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    image_bytes = await file.read()
    result = await detection_service.process_scan(
        db=db,
        user_id=current_user.id,
        image_bytes=image_bytes,
        original_filename=file.filename or "leaf_scan.jpg",
        farm_id=farm_id,
        field_id=field_id,
        crop_id=crop_id
    )
    return success_response(data=result.dict(), message="Crop scan and diagnosis completed successfully")

@router.get("", response_model=dict)
async def list_detections(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    crop_id: Optional[str] = Query(None),
    farm_id: Optional[str] = Query(None),
    field_id: Optional[str] = Query(None),
    detection_type: Optional[str] = Query(None),
    severity: Optional[str] = Query(None),
    risk_level: Optional[str] = Query(None),
    expert_verified: Optional[bool] = Query(None),
    search: Optional[str] = Query(None),
    sort_by: Optional[str] = Query("newest"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    # Only filter by user_id if farmer; experts & admins can view all or own
    user_filter = current_user.id if current_user.role == "FARMER" else None

    params = DetectionFilterParams(
        page=page,
        page_size=page_size,
        crop_id=crop_id,
        farm_id=farm_id,
        field_id=field_id,
        detection_type=detection_type,
        severity=severity,
        risk_level=risk_level,
        expert_verified=expert_verified,
        search=search,
        sort_by=sort_by
    )

    result = await detection_service.list_detections(db, user_filter, params)
    return success_response(data=result.dict())

@router.get("/{detection_id}", response_model=dict)
async def get_detection(
    detection_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    result = await detection_service.get_detection_detail(db, detection_id, user_id=current_user.id)
    return success_response(data=result.dict())
